arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorHolder.java

```diff
@@ -140,12 +141,18 @@ public static class ConstantVectorHolder<T> extends VectorHolder {
     private final int numRows;
 
     public ConstantVectorHolder(int numRows) {
+      super(new NullVector("_dummy_", numRows), null, new NullabilityHolder(numRows));
+      nullabilityHolder().setNulls(0, numRows);
       this.numRows = numRows;
       this.constantValue = null;
     }
 
     public ConstantVectorHolder(Types.NestedField icebergField, int numRows, T constantValue) {
-      super(icebergField);
+      super(
+          new NullVector(icebergField.name(), numRows),
```

Comment:
@nastra, @amogh-jahagirdar  This change causes the Spark tests the fail. I've analyzed the failure and concluded that this change, as is, assumes that ConstantVectorHolder will only ever hold a null value. This patch fixes the issue, but it feels a bit tacky to me. What do you think?

```
Subject: [PATCH] Changes
---
Index: arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorHolder.java
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorHolder.java b/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorHolder.java
--- a/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorHolder.java	(revision 1a3896bba82e1e92142b237e1cd9332d67171b19)
+++ b/arrow/src/main/java/org/apache/iceberg/arrow/vectorized/VectorHolder.java	(date 1727193972380)
@@ -149,10 +149,12 @@
 
     public ConstantVectorHolder(Types.NestedField icebergField, int numRows, T constantValue) {
       super(
-          new NullVector(icebergField.name(), numRows),
+          (null == constantValue) ? new NullVector(icebergField.name(), numRows) : null,
           icebergField,
           new NullabilityHolder(numRows));
-      nullabilityHolder().setNulls(0, numRows);
+      if (null == constantValue) {
+        nullabilityHolder().setNulls(0, numRows);
+      }
       this.numRows = numRows;
       this.constantValue = constantValue;
     }
```

Reply:
@slessard what was the issue with the previous patch I proposed to you? I don't recall spark tests failing there. Also what was the Spark failure?

Reply:
Below is the link to the Spark failure. I'll do a more detailed writeup of the issue in a few minutes. 
https://github.com/apache/iceberg/actions/runs/10964734930/job/30505018420.

Reply:
@nastra  There are many test classes that now contain failures. Here's a good example of just one of the Spark test classes that fails. Most of the tests in TestSparkDataWrite are failing.

```
org.opentest4j.AssertionFailedError: [Result rows should match] 
expected: [{"id"=1,"data"="a"}, {"id"=2,"data"="b"}, {"id"=3,"data"="c"}]
 but was: [{"id"=1,"data"="null"}, {"id"=2,"data"="null"}, {"id"=3,"data"="null"}]
	at org.apache.iceberg.spark.source.TestSparkDataWrite.testBasicWrite(TestSparkDataWrite.java:145)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	...
```

Reply:
@nastra  OK, Now I see what's going on. The Spark tests were failing before I added a the VectorHolderTest class. I added that class because I realized that there are three different ways to end up with a dummy holder. I ran that test class against the main branch and all the tests passed. When I ran that test class against my branch some of the tests for the isDummy method failed. I then changed the isDummy method to correctly handle the three cases

1.  The dummy holder is a DeletedVectorHolderInstance via the deletedVectorHolder() method.
2. The dummy holder is a ConstantVectorHolder created via the dummyHolder() method, i.e. vector has no assigned value
3. the dummy holder is a Constant vector holder created via the constantHolder() method, i.e. vector has an assigned value

For DeletedVectorHolder its `vector` instance variable is always `null`.
For ConstantVectorHolder, regardless of whether its constant value is `"a"` or `null`, its `vector` instance variable is always a `NullVector` instance.

I updated the `isDummy()` method to handle those two cases. This made the three unit tests related to the isDummy method pass.

Fixing the isDummy method seems to have fixed the Spark tests.

The only remaining issue (possibly minor enough to ignore) is that ConstantVectorHolder, regardless of whether its constant value is `"a"` or `null`, its `vector` instance variable is always a `NullVector` instance. This seems a bit misleading, especially when the constant value is `"a"`. My suggested patch above fixes this issue to make ConstantVectorHolder's `vector` have a `null` value when its constant value is `"a"`. Is this worth fixing? If yes, then my suggested patch is needed. If no, then my suggested patch is not needed.

Reply:
I think let's apply the patch you suggested above

Reply:
patch has been applied